app/javascript/packages/document-capture/components/full-screen.jsx

```diff
@@ -1,81 +1,139 @@
+ * @prop {RequestCloseCallback=} onRequestClose Callback invoked when user initiates close intent.
+ * @prop {string} label Accessible label for modal.
+ * @prop {ReactNode} children Child elements.
  */
-let activeInstances = 0;
 
 /**
- * @param {FullScreenProps} props Props object.
+ * @param {React.MutableRefObject<HTMLElement?>} containerRef
+ * @param {RequestCloseCallback} onRequestClose
  */
-function FullScreen({ onRequestClose = () => {}, children }) {
-  const { t } = useI18n();
-  const { getAssetPath } = useAsset();
-  const trapRef = useRef(/** @type {?import('focus-trap').FocusTrap} */ (null));
+function useFocusTrap(containerRef, onRequestClose) {
+  const trapRef = useRef(/** @type {FocusTrap?} */ (null));
   const onRequestCloseRef = useRef(onRequestClose);
+
   useEffect(() => {
     // Since the focus trap is only initialized once, but the callback could
     // be changed, ensure that the current reference is kept as a mutable value
     // to reference in the deactivation.
     onRequestCloseRef.current = onRequestClose;
   }, [onRequestClose]);
 
-  const setFocusTrapRef = useCallback((node) => {
-    if (trapRef.current) {
-      trapRef.current.deactivate();
-    }
-
-    if (node) {
-      trapRef.current = createFocusTrap(node, {
+  useEffect(() => {
+    if (containerRef.current) {
+      trapRef.current = createFocusTrap(containerRef.current, {
         onDeactivate: () => onRequestCloseRef.current(),
         clickOutsideDeactivates: true,
       });
 
       trapRef.current.activate();
     }
-  }, []);
-
-  useEffect(() => {
-    if (activeInstances++ === 0) {
-      document.body.classList.add('has-full-screen-overlay');
-      document.documentElement.classList.add('has-full-screen-overlay');
-    }
 
     return () => {
-      if (--activeInstances === 0) {
-        document.body.classList.remove('has-full-screen-overlay');
-        document.documentElement.classList.remove('has-full-screen-overlay');
-      }
+      trapRef.current?.deactivate();
     };
   }, []);
+}
+
+const useBodyClass = (() => {
+  /**
+   * @type {WeakMap<any, number>}
+   */
+  const activeInstancesByComponent = new WeakMap();
 
   return (
-    <div ref={setFocusTrapRef} aria-modal="true" className="full-screen bg-white">
+    /**
+     * @param {string} className Class name to add to body element
+     * @param {React.ComponentType<any>} Component React component definition
+     */
+    (className, Component) => {
+      /**
+       * Increments the number of active instances for the current component by the given amount,
+       * adding or removing the body class for the first and last instance respectively.
+       *
+       * @param {number} amount
+       */
+      function incrementActiveInstances(amount) {
+        const activeInstances = activeInstancesByComponent.get(Component) || 0;
+        const nextActiveInstances = activeInstances + amount;
+
+        if (nextActiveInstances === 1 || nextActiveInstances === 0) {
```

Comment:
On reflection, I see this condition is rather naive. `nextActiveInstances` could be `1` due to unmounting one of two active instances

```
activeInstances = 2
amount = -1
nextActiveInstances = activeInstances + amount
nextActiveInstances === 1
```

It won't hurt because the effect is the same, but we should want to try to minimize unnecessary calls to `classList#toggle` if we can, I think.

Reply:
> On reflection, I see this condition is rather naive. `nextActiveInstances` could be `1` due to unmounting one of two active instances

This was addressed in 44029b1dbab8538998e992b69c1b98cb89cc1df1. New logic considers changing from zero to non-zero, and vice versa:

```js
    if (!activeInstances && nextActiveInstances) {
      document.body.classList.add(className);
    } else if (activeInstances && !nextActiveInstances) {
      document.body.classList.remove(className);
    }
```

Note: This also undoes the work from #4259. I wasn't able to reproduce the original issue, but I'll plan to do more testing just in case.